packages/plugin-js-packages/src/lib/runner/audit/constants.ts

```diff
@@ -28,9 +30,41 @@ export const normalizeAuditMapper: Record<
   npm: npmToAuditResult,
   'yarn-classic': yarnv1ToAuditResult,
   'yarn-modern': yarnv2ToAuditResult,
-  pnpm: () => {
-    throw new Error('PNPM audit is not supported yet.');
-  },
+  pnpm: pnpmToAuditResult,
+};
+
+const filterNpmAuditResults = (
+  results: Record<DependencyGroup, AuditResult>,
+) => ({
+  prod: results.prod,
+  dev: filterAuditResult(results.dev, 'name', results.prod),
+  optional: filterAuditResult(results.optional, 'name', results.prod),
+});
+
+const filterPnpmAuditResults = (
+  results: Record<DependencyGroup, AuditResult>,
+) => ({
+  prod: results.prod,
+  dev: results.dev,
+  optional: filterAuditResult(
+    filterAuditResult(results.optional, 'id', results.prod),
+    'id',
+    results.dev,
+  ),
+});
+
+export const postProcessingAuditMapper: Record<
+  PackageManager,
+  (
+    result: Record<DependencyGroup, AuditResult>,
+  ) => Record<DependencyGroup, AuditResult>
+> = {
+  npm: filterNpmAuditResults, // prod dependencies need to be filtered out manually since v10
+  /* eslint-disable @typescript-eslint/naming-convention */
+  'yarn-classic': results => results,
+  'yarn-modern': results => results,
+  /* eslint-enable @typescript-eslint/naming-convention */
+  pnpm: filterPnpmAuditResults, // optional dependencies don't have an exclusive option so they need duplicates filtered out
 };
```

Comment:
I'm not fully aware of the full implementation so please ignore the question if not fitting.

Not sure if it is a good approach to group different implementations of the same thing within one thing. e.g. npm and pnpm in the same object.

The buzzword that pops up here for me is polymorphism. You basically have a shared API e.g. `filterResults` and a list that maintains the different implementations of that interface. then you call always the same API and get rid of branching and other code clutter.

Reply:
Good insight, thank you!
This will, however, require a bigger refactor so I will address this in a follow-up PR.

Reply:
I agree with both the proposed refactor and doing it in a different PR :+1:

Reply:
Included as part of #597